conmon/conmon.c

```diff
@@ -116,6 +116,65 @@ int set_k8s_timestamp(char *buf, ssize_t buflen, const char *stream_type)
 	return 0;
 }
 
+
+/*
+ * splits buf into lines and inserts timestamps at the beginning of
+ * the line written to logfd.
+ */
+int write_with_timestamps(int logfd, const char *buf, ssize_t buflen)
+{
+	#define TSBUFLEN 34
+	char tsbuf[TSBUFLEN];
+	static bool last_buf_ended_with_newline = TRUE;
+
+	g_auto(GStrv) lines = g_strsplit(buf, "\n", -1);
+	ssize_t num_lines = g_strv_length(lines);
+	for (ssize_t i = 0; i < num_lines; i++)
+	{
+		const char *line = lines[i];
+
+		ninfo("Processing line: %ld, %s", i, line);
+
+		/* Skip last line if it is empty */
+		if (i == (num_lines - 1) && buf[buflen-1] == '\n' && !strcmp("", line)) {
+			ninfo("Skipping last line");
+			break;
+		}
+
+		/*
+		 * Only add timestamps for first line if last buffer's last
+		 * line ended with newline. Add it for all other lines.
+		 */
+		if (i != 0 || (i == 0 && last_buf_ended_with_newline)) {
+			ninfo("Adding timestamp");
+			int rc = set_k8s_timestamp(tsbuf, TSBUFLEN, "stdout");
```

Comment:
`snprintf` in `set_k8s_timestamp` is adding a `\0` char at the end of the buf. When the log is read by k8s the hex dump is as follow:
```
002f62696e2f73683a2063616e277420637265617465202f66696c653a20526561642d6f6e6c792066696c652073797374656d0a
```
notice the `00` at the beginning. Which makes this test fail with:
```
    it should not write to root filesystem [Conformance] [It]
    /home/amurdaca/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:154

    Timed out after 60.000s.
    Expected
        <string>: /bin/sh: can't create /file: Read-only file system

    to equal
        <string>: /bin/sh: can't create /file: Read-only file system
```
even if strings appear to be the same.

The thing should be that, when we write the timestamp line with the log, we should `snprintf` with the whole log line, this patch does that but it's ugly I know :):
```diff
diff --git a/conmon/conmon.c b/conmon/conmon.c
index f69720f..8c4fdc7 100644
--- a/conmon/conmon.c
+++ b/conmon/conmon.c
@@ -93,7 +93,7 @@ static GOptionEntry entries[] =
   { NULL }
 };
 
-int set_k8s_timestamp(char *buf, ssize_t buflen, const char *stream_type)
+int set_k8s_timestamp(int logfd, char *buf, ssize_t buflen, const char *stream_type, const char *log)
 {
 	time_t now = time(NULL);
 	struct tm *tm;
@@ -108,10 +108,17 @@ int set_k8s_timestamp(char *buf, ssize_t buflen, const char *stream_type)
 		off_sign = '-';
 		off = -off;
 	}
-	snprintf(buf, buflen, "%d-%02d-%02dT%02d:%02d:%02d%c%02d:%02d %s ",
+	snprintf(buf, buflen, "%d-%02d-%02dT%02d:%02d:%02d%c%02d:%02d %s %s",
 		tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
 		tm->tm_hour, tm->tm_min, tm->tm_sec,
-		off_sign, off / 3600, off % 3600, stream_type);
+		off_sign, off / 3600, off % 3600, stream_type, log);
+
+	char newbuf[buflen-1];
+	strncpy(newbuf, buf, buflen-1);
+
+	if (write(logfd, newbuf, buflen-1) != buflen) {
+		nwarn("partial/failed write ts (logFd)");
+	}
 
 	return 0;
 }
@@ -124,7 +131,6 @@ int set_k8s_timestamp(char *buf, ssize_t buflen, const char *stream_type)
 int write_with_timestamps(int logfd, const char *buf, ssize_t buflen)
 {
 	#define TSBUFLEN 34
-	char tsbuf[TSBUFLEN];
 	static bool last_buf_ended_with_newline = TRUE;
 
 	g_auto(GStrv) lines = g_strsplit(buf, "\n", -1);
@@ -147,21 +153,24 @@ int write_with_timestamps(int logfd, const char *buf, ssize_t buflen)
 		 */
 		if (i != 0 || (i == 0 && last_buf_ended_with_newline)) {
 			ninfo("Adding timestamp");
-			int rc = set_k8s_timestamp(tsbuf, TSBUFLEN, "stdout");
+			ssize_t len = strlen(line);
+			char tsbuf[TSBUFLEN+len];
+			ninfo("fuck %ld", TSBUFLEN+len);
+			int rc = set_k8s_timestamp(logfd, tsbuf, TSBUFLEN+len, "stdout", line);
 			if (rc < 0) {
 				nwarn("failed to set timestamp");
 			} else {
-				if (write(logfd, tsbuf, TSBUFLEN) != TSBUFLEN) {
-					nwarn("partial/failed write ts (logFd)");
-				}
+				/*if (write(logfd, tsbuf, TSBUFLEN+len) != TSBUFLEN+len) {*/
+					/*nwarn("partial/failed write ts (logFd)");*/
+				/*}*/
+			}
+		} else {
+			/* Log output to logfd. */
+			ssize_t len = strlen(line);
+			if (write(logfd, line, len) != len) {
+				nwarn("partial/failed write (logFd)");
+				return -1;
 			}
-		}
-
-		/* Log output to logfd. */
-		ssize_t len = strlen(line);
-		if (write(logfd, line, len) != len) {
-			nwarn("partial/failed write (logFd)");
-			return -1;
 		}
 		/* Write the line ending */
 		if (write(logfd, "\n", 1) != 1) {
@@ -483,7 +492,7 @@ int main(int argc, char *argv[])
 					if (num_read <= 0)
 						goto out;
 
-					buf[num_read] = '\0';
+					/*buf[num_read] = '\0';*/
 					ninfo("read a chunk: (fd=%d) '%s'", mfd, buf);
 
 					/* Insert CRI mandated timestamps in the buffer for each line */
```
This patch makes the test above pass. The patch is ugly but the thing is we should not play with null-terminated strings when printing to `logfd`, or when read back we have issues.